File: packages/frontend/e2e/study-records.spec.ts
import { test, expect, type Page } from "@playwright/test";

/**
 * 学習記録画面（/study-records）の代表導線を検証する E2E。
 * GraphQL はモックし、「カレンダー表示 → 日付選択で詳細 → 月送り」と
 * streak / 週ストリーク・ダッシュボード（直近一覧）の表示を確認する。
 * カレンダーは常に当月から始まるため、テストデータは実行日基準で組み立てる。
 */

// ローカル日付を "YYYY-MM-DD" に整形（ISO 変換のタイムゾーンずれを避ける）。
function formatDate(d: Date) {
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, "0");
  const day = String(d.getDate()).padStart(2, "0");
  return `${y}-${m}-${day}`;
}

const today = new Date();
const year = today.getFullYear();
const month = today.getMonth() + 1;
const todayStr = formatDate(today);

// studyCount はカレンダーの日付数字（1〜31）と衝突しない値にして、
// バッジのテキストを exact 一致で特定できるようにする。
const studyRecord = {
  id: "1",
  studyDate: todayStr,
  studyCount: 99,
  studyDetails: [
    {
      id: "10",
      title: "TOEIC part1",
      rate: 80,
      totalCount: 10,
      correctCount: 8,
      __typename: "StudyDetail",
    },
  ],
  __typename: "StudyRecord",
};

async function mockGraphql(page: Page) {
  await page.route("**/graphql", async (route) => {
    const body = route.request().postDataJSON() as {
      operationName?: string;
      variables?: { year?: number; month?: number };
    };
    const op = body?.operationName;

    if (op === "Me") {
      await route.fulfill({
        contentType: "application/json",
        body: JSON.stringify({
          data: {
            me: {
              id: "1",
              name: "テスト太郎",
              email: "taro@example.com",
              role: "user",
              streak: 3,
              wordsCount: 0,
              __typename: "User",
            },
          },
        }),
      });
      return;
    }

    if (op === "StudyRecords") {
      // 当月だけ記録あり。月送り後（別の年月）は空を返す。
      const isCurrentMonth =
        body.variables?.year === year && body.variables?.month === month;
      await route.fulfill({
        contentType: "application/json",
        body: JSON.stringify({
          data: { studyRecords: isCurrentMonth ? [studyRecord] : [] },
        }),
      });
      return;
    }

    if (op === "StudyRecordsWeek") {
      await route.fulfill({
        contentType: "application/json",
        body: JSON.stringify({ data: { studyRecordsWeek: [studyRecord] } }),
      });
      return;
    }

    if (op === "StudyRecordsRecent") {
      await route.fulfill({
        contentType: "application/json",
        body: JSON.stringify({ data: { studyRecordsRecent: [studyRecord] } }),
      });
      return;
    }

    await route.fulfill({
      contentType: "application/json",
      body: JSON.stringify({ data: { me: null } }),
    });
  });
}

test("カレンダーと streak が表示され、日付選択で詳細が出る", async ({
  page,
}) => {
  await mockGraphql(page);
  await page.goto("/study-records");

  // 見出し + streak（me.streak = 3 日）
  await expect(page.getByRole("heading", { name: "学習記録" })).toBeVisible();
  await expect(page.getByText("連続学習3日")).toBeVisible();

  // 当月カレンダーに学習数バッジ（studyCount）が出る
  await expect(
    page.getByRole("heading", { name: `${year}年${month}月` }),
  ).toBeVisible();
  await expect(page.getByText("99", { exact: true })).toBeVisible();

  // 学習した日を選ぶと study_details（タイトル / 正答数 / 正答率）が出る
  await page
    .getByRole("button", { name: `${month}月${today.getDate()}日` })
    .click();
  await expect(page.getByText("TOEIC part1")).toBeVisible();
  await expect(page.getByText("8/10")).toBeVisible();
  await expect(page.getByText("80%")).toBeVisible();
});

test("月送りで前月・翌月に切り替わる", async ({ page }) => {
  await mockGraphql(page);
  await page.goto("/study-records");

  await expect(
    page.getByRole("heading", { name: `${year}年${month}月` }),
  ).toBeVisible();

  // 前の月へ（記録なしの月はバッジも消える）
  const prev = new Date(year, month - 1 - 1, 1);
  await page.getByRole("button", { name: "前の月" }).click();
  await expect(
    page.getByRole("heading", {
      name: `${prev.getFullYear()}年${prev.getMonth() + 1}月`,
    }),
  ).toBeVisible();
  await expect(page.getByText("99", { exact: true })).not.toBeVisible();

  // 次の月へ 2 回で翌月まで戻る
  await page.getByRole("button", { name: "次の月" }).click();
  await page.getByRole("button", { name: "次の月" }).click();
  const next = new Date(year, month - 1 + 1, 1);
  await expect(
    page.getByRole("heading", {
      name: `${next.getFullYear()}年${next.getMonth() + 1}月`,
    }),
  ).toBeVisible();
});

test("ダッシュボードタブで直近の学習記録一覧が出る", async ({ page }) => {
  await mockGraphql(page);
  await page.goto("/study-records");

  await page.getByRole("button", { name: "ダッシュボード" }).click();
  await expect(
    page.getByText("最近の学習記録一覧（最新30件）"),
  ).toBeVisible();
  await expect(page.getByText("TOEIC part1")).toBeVisible();
  await expect(page.getByText("words")).toBeVisible();
});

File: packages/backend/spec/requests/graphql/study_records_spec.rb
require "rails_helper"

RSpec.describe Resolvers::StudyRecords do
  let(:query) do
    <<~GQL
      query($year: Int!, $month: Int!) {
        studyRecords(year: $year, month: $month) {
          id
          studyDate
          studyCount
          studyDetails { id title rate totalCount correctCount }
        }
      }
    GQL
  end

  let!(:user) { create(:user) }

  describe "正常系" do
    it "指定月の記録を study_details ごと日付昇順で返す" do
      later = create(:study_record, user: user, study_date: Date.new(2026, 7, 20), study_count: 5)
      earlier = create(:study_record, user: user, study_date: Date.new(2026, 7, 3), study_count: 10)
      create(:study_detail, study_record: earlier,
        title: "TOEIC part1", rate: 80, total_count: 10, correct_count: 8)

      result = execute_graphql(query, variables: { year: 2026, month: 7 },
        context: { current_user: user })
      records = result.dig("data", "studyRecords")

      expect(records.map { |r| r["studyDate"] }).to eq([ "2026-07-03", "2026-07-20" ])
      expect(records.first["studyCount"]).to eq(10)
      expect(records.first["studyDetails"]).to eq([ {
        "id" => earlier.study_details.first.id.to_s,
        "title" => "TOEIC part1", "rate" => 80, "totalCount" => 10, "correctCount" => 8
      } ])
      expect(records.second["id"]).to eq(later.id.to_s)
    end

    it "月境界：月初・月末は含み、前月末・翌月頭は含まない" do
      create(:study_record, user: user, study_date: Date.new(2026, 6, 30))
      create(:study_record, user: user, study_date: Date.new(2026, 7, 1))
      create(:study_record, user: user, study_date: Date.new(2026, 7, 31))
      create(:study_record, user: user, study_date: Date.new(2026, 8, 1))

      result = execute_graphql(query, variables: { year: 2026, month: 7 },
        context: { current_user: user })
      dates = result.dig("data", "studyRecords").map { |r| r["studyDate"] }

      expect(dates).to eq([ "2026-07-01", "2026-07-31" ])
    end

    it "記録が無い月は空配列を返す" do
      result = execute_graphql(query, variables: { year: 2026, month: 1 },
        context: { current_user: user })

      expect(result.dig("data", "studyRecords")).to eq([])
    end
  end

  describe "認可・異常系" do
    it "他人の記録は含めない（current_user スコープ）" do
      create(:study_record, user: create(:user), study_date: Date.new(2026, 7, 5))

      result = execute_graphql(query, variables: { year: 2026, month: 7 },
        context: { current_user: user })

      expect(result.dig("data", "studyRecords")).to eq([])
    end

    it "不正な月は BAD_REQUEST" do
      result = execute_graphql(query, variables: { year: 2026, month: 13 },
        context: { current_user: user })

      expect(result.dig("data", "studyRecords")).to be_nil
      expect(result.dig("errors", 0, "extensions", "code")).to eq("BAD_REQUEST")
    end

    it "未認証は UNAUTHORIZED" do
      result = execute_graphql(query, variables: { year: 2026, month: 7 })

      expect(result.dig("data", "studyRecords")).to be_nil
      expect(result.dig("errors", 0, "extensions", "code")).to eq("UNAUTHORIZED")
    end
  end
end
